migrations/kdvh/dump.go

```diff
@@ -0,0 +1,244 @@
+package kdvh
+
+import (
+	"database/sql"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	_ "github.com/jackc/pgx/v5/stdlib"
+
+	"migrate/utils"
+)
+
+type DumpConfig struct {
+	BaseDir     string   `short:"p" long:"path" default:"./dumps/kdvh" description:"Location the dumped data will be stored in"`
+	TablesCmd   string   `short:"t" long:"table" default:"" description:"Optional comma separated list of table names. By default all available tables are processed"`
+	StationsCmd string   `short:"s" long:"stnr" default:"" description:"Optional comma separated list of stations IDs. By default all station IDs are processed"`
+	ElementsCmd string   `short:"e" long:"elem" default:"" description:"Optional comma separated list of element codes. By default all element codes are processed"`
+	Overwrite   bool     `long:"overwrite" description:"Overwrite any existing dumped files"`
+	Email       []string `long:"email" description:"Optional email address used to notify if the program crashed"`
+
+	Tables   []string
+	Stations []string
+	Elements []string
+}
+
+func (config *DumpConfig) setup() {
+	if config.TablesCmd != "" {
+		config.Tables = strings.Split(config.TablesCmd, ",")
+	}
+	if config.StationsCmd != "" {
+		config.Stations = strings.Split(config.StationsCmd, ",")
+	}
+	if config.ElementsCmd != "" {
+		config.Elements = strings.Split(config.ElementsCmd, ",")
+	}
+}
+
+func (config *DumpConfig) Execute([]string) error {
+	config.setup()
+
+	conn, err := sql.Open("pgx", os.Getenv("KDVH_PROXY_CONN"))
+	if err != nil {
+		slog.Error(err.Error())
+		return nil
+	}
+
+	for _, table := range KDVH {
+		if config.Tables != nil && !slices.Contains(config.Tables, table.TableName) {
+			continue
+		}
+		table.Dump(conn, config)
+	}
+
+	return nil
+}
+
+func (table *Table) Dump(conn *sql.DB, config *DumpConfig) {
+	defer utils.SendEmailOnPanic(fmt.Sprintf("%s dump", table.TableName), config.Email)
+
+	table.Path = filepath.Join(config.BaseDir, table.Path)
+	if _, err := os.ReadDir(table.Path); err == nil && !config.Overwrite {
+		slog.Info(fmt.Sprint("Skipping data dump of ", table.TableName, " because dumped folder already exists"))
+		return
+	}
+
+	utils.SetLogFile(table.TableName, "dump")
+	elements, err := table.getElements(conn, config)
+	if err != nil {
+		return
+	}
+
+	bar := utils.NewBar(len(elements), table.TableName)
+
+	// TODO: should be safe to spawn goroutines/waitgroup here with connection pool?
+	bar.RenderBlank()
+	for _, element := range elements {
+		table.dumpElement(element, conn, config)
+		bar.Add(1)
+	}
+}
+
+// TODO: maybe we don't do this? Or can we use pgdump/copy?
+// The problem is that there are no indices on the tables, that's why the queries are super slow
```

Comment:
Maybe I'm wrong (I need to review my indices :sweat_smile:), but I'm using WHERE to exclude NULL observations (for a specific element column) and I also specify a single `stnr`, wouldn't that be faster to scan if there was an index on the columns?

Reply:
Ahh I see what you mean. That would require an index on every single element though, which I think would cause write perf problems for KDVH. As it is they're already struggling with write throughput.

Specifying a single station number should be helpful though (I think I hadn't realised you were dumping station by station when I wrote that comment) I'm shocked that they don't have at least in index on the station number column?!

Reply:
Maybe they have them (directly in Oracle, which are then not reflected in the proxy) and I'm simply wrong. I was also wrong in thinking that indeces would affect distinct queries, you learn something new everyday :)